compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,6 +286,212 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
     }
   }
 
+  /**
+   * Create a threadsafe lazy accessor equivalent to the following code:
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * @tailrec def x: A =
+   *   _x match
+   *   case null =>
+   *     if CAS(_x, null, Evaluating) then
+   *       var result: AnyRef = null // here, we need `AnyRef` to possibly assign `NullValue`
+   *       try
+   *         result = rhs
+   *         nullable = null // if the field is nullable; see `CollectNullableFields`
+   *         if result == null then result = NullValue // drop if A is non-nullable
+   *       finally
+   *         if !CAS(_x, Evaluating, result) then
+   *           val lock = _x.asInstanceOf[Waiting]
+   *           CAS(_x, lock, result)
+   *           lock.release()
+   *     x
+   *   case Evaluating =>
+   *     CAS(_x, Evaluating, new Waiting)
+   *     x
+   *   case current: Waiting =>
+   *     current.awaitRelease()
+   *     x
+   *   case NullValue => null
+   *   case current: A =>
+   *      current
+   * ```
+   * Where `Evaluating` and `NullValue` are represented by `object`s and `Waiting` by a class that
+   * allows awaiting the completion of the evaluation. Note that since tail-recursive
+   * functions are transformed *before* lazy-vals, this implementation directly implements
+   * the resulting loop. `PatternMatcher` coming before `LazyVals`, the pattern matching block
+   * is implemented using if-s. That is:
+   * 
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * def x: A =
+   *   while true do
+   *     val current: AnyRef = _x
+   *     if current == null then
+   *       if CAS(_x, null, Evaluating) then
+   *         var result: AnyRef = null
+   *         try
+   *           result = rhs
+   *           nullable = null
+   *           if result == null then result = NullValue
+   *         finally
+   *           if !CAS(_x, Evaluating, result) then
+   *             val lock = _x.asInstanceOf[Waiting]
+   *             CAS(_x, lock, result)
+   *             lock.release()
+   *     else
+   *       if current.isInstanceOf[Evaluating] then
+   *         CAS(current, Evaluating, new Waiting)
+   *       else if current.isInstanceOf[NullValue] then
+   *         return null
+   *       else if current.isInstanceOf[Waiting] then
+   *         current.asInstanceOf[Waiting].await()
+   *       else
+   *         return current.asInstanceOf[A]
+   *   end while
```

Comment:
This can help, but care must be taken to only do that when the (erased) type `A` of the field is none of the supertypes of `Waiting`. That means `Object`, obviously, but also `CountDownLatch`.